README.md
# holistic-image

> Holism is the idea that various **systems should be viewed as wholes**, not merely as a collection of parts.

Build-time Automatic image transformation and Holistic management

- 🍊 uses [squoosh](https://github.com/GoogleChromeLabs/squoosh/tree/dev/libsquoosh) to derive jpg, webp, avif from your
  sources
- 📦 hides implementation details behind Webpack
- 🤖 on demand file creation, and CLI utils to verify integrity
- ⚛️ optional React implementation

# Structure

This is a _convention over configuration_ library, and all you need is to follow our convention

Having ➡️

```
├── image@2x.holistic.png
```

️Will produce ⬇️

```
├── image@2x.holistic.png
├── .holistic (you can hide this directory)
│   └─ image@2x
│      ├─ derived.image@1x.jpg
│      ├─ derived.image@1x.webp
│      ├─ derived.image@1x.avif
│      ├─ derived.image@2x.jpg
│      ├─ derived.image@2x.webp
│      ├─ derived.image@2x.avif
│      └─ derived.image@2x.meta.js
```

The same principle will be applied during the import - instead of importing `image@2x.holistic.png` you will get a
pointer to all files below

> Note: holistic-image does not producing `png` output (configurable) as the end user is expected to use `webp` or `avif`

# Usage

## Step 1 - derive files

`holistic-image` is looking for files named accordingly - `image.holistic.png`(or jpg) and **derives**
the "missing" ones - optimized `jpg`, `webp` and `avif`

If the source file named as `image@2x.jpg`(Figma standard), then `@1x` version will be generated automatically

### How to use

- via Webpack loader Just use webpack loader with `autogenerate` option enabled (default)
- via API

```ts
// generate-images.js
import {deriveHolisticImages} from "holistic-image/api";

deriveHolisticImages(
  /root folder*/
process.argv[2],
  /*mask*/ process.argv[3],
// /*optional*/ squoosh ecoders with options
)
```

And then in `package.json`

```
// package.json
 "autogen:images": "yarn generate-images.js $INIT_CWD 'src/**/*'",
```

- via CLI

```
// package.json
"autogen:images":"holistic-image derive  $INIT_CWD 'src/**/*'"
"validate:images":"holistic-image validate  $INIT_CWD 'src/**/*'"
```

## Step 2 - configure webpack to process images

- Optimized config, will remove originals from the final bundle

```ts
import { holisticImage } from 'holistic-image/webpack';

webpack.config = {
  module: {
    rules: {
      oneOf: [holisticImage, yourFileLoader],
      // .. rest of your config
    },
  },
};
```

- automatic image generation will be enabled if `process.env.NODE_ENV` is set to `development`
- to fine control settings use:

  - `import { holisticImagePresetFactory } from 'holistic-image/webpack';`
  - `import { holisticImageLoader } from 'holistic-image/webpack';`

- Easy config (for storybook for example), everything will work as well

```ts
import { holisticImage } from 'holistic-image/webpack';

webpack.config = {
  module: {
    rules: {
      holisticImage,
      yourFileLoader,
      // .. and rest of your config
    },
  },
};
```

## Step 3 - use

```ts
import image from './image.holistic.jpg';
// ^ not an image, but HolisticalImageDefinition
image = {
  base: [1x, 2x],
  webp: [1x, 2x],
  avif: [1x, 2x],
  [META]: {width, height}
}
```

### Build in React component

```tsx
import { Image } from 'holistical-image/react';
import image from './image.holistic.jpg';
import imageXS from './imageXS.holistic.jpg';

<Image src={image} media={{ 'max-width: 600px': imageXS }} />;
// 👇 6-12 images generated, the right picked, completely transparent
```

## TypeScript integration

While this library provides `d.ts` for the file extension it can be more beneficial to provide your own ones, as you did
for `.jpg` and other static asses already

```ts
declare module '*.holistic.jpg' {
  import type { HolisticalImageDefinition } from 'holistic-image';

  const content: HolisticalImageDefinition;
  export default content;
}
```

# Configuration

Configuration is possible in two modes:

- full control via API
- config-file based (via [cosmic-config](https://github.com/davidtheclark/cosmiconfig))

<details>
<summary>Example configuration:</summary>

> `.holistic-imagerc.yaml` (can be json or js as well)

```yml
# derived from https://github.com/GoogleChromeLabs/squoosh/blob/61de471e52147ecdc8ff674f3fcd3bbf69bb214a/libsquoosh/src/codecs.ts
---
jpg:
  use: mozjpeg
  # with default 75
  options:
    - quality: 80 # for scale 1x
    - quality: 70 # for scale 2x
webp:
  use: webp
  # with default 75
  options:
    - quality: 85
      method: 6
avif:
  use: avif
  # with default 33
  options:
    - cqLevel: 20
      effort: 5
    - cqLevel: 28
      effort: 5
```

</details>

# Hiding .holistic output files

folders starting from `.` already expected to be hidden for IDE, but keep in mind - derived files are **expected to be
commited**.

## WebStorm/IDEA

You can use [idea-exclude](https://github.com/theKashey/idea-exclude) to automaticaly configure Idea-based solutions
to _exclude_ these folders

- run `idea-exclude holistic-images "src/**/.holistic"`

## VCS

```
"files.exclude": {
    "**/.holistic": true
}
```

# See also

- [imagemin](https://github.com/imagemin/imagemin) (unmaintaned) the same _defiving_ mechanics, with no further
  management
- [image-webpack-loader](https://github.com/tcoopman/image-webpack-loader) not creates, but optimizes existing images
- [nextjs/image](https://nextjs.org/docs/api-reference/next/image) serves optimized image via CDN transformation

# License

MIT


CHANGELOG.md
## [1.2.3](https://github.com/theKashey/holistic-image/compare/v1.2.2...v1.2.3) (2021-07-17)

### Bug Fixes

- fine type compression quality settings between 1x and 2x ([8e41df1](https://github.com/theKashey/holistic-image/commit/8e41df197d127bafe8dbb42bab3927f7143758d6))

## [1.2.2](https://github.com/theKashey/holistic-image/compare/v1.2.1...v1.2.2) (2021-07-15)

### Bug Fixes

- correct behavior for 1x images ([df0955d](https://github.com/theKashey/holistic-image/commit/df0955d26e6895aad3dafcea4f758b718031ca65))

## [1.2.1](https://github.com/theKashey/holistic-image/compare/v1.2.0...v1.2.1) (2021-07-15)

### Bug Fixes

- ease mtime comparison, implements [#13](https://github.com/theKashey/holistic-image/issues/13) ([7885863](https://github.com/theKashey/holistic-image/commit/78858634540abf4c8e4226a96d2b9adede827b9f))

# [1.2.0](https://github.com/theKashey/holistic-image/compare/v1.1.6...v1.2.0) (2021-07-14)

## [1.1.6](https://github.com/theKashey/holistic-image/compare/v1.1.5...v1.1.6) (2021-07-14)

## [1.1.5](https://github.com/theKashey/holistic-image/compare/v1.1.4...v1.1.5) (2021-07-13)

## [1.1.4](https://github.com/theKashey/holistic-image/compare/v1.1.3...v1.1.4) (2021-07-13)

## [1.1.3](https://github.com/theKashey/holistic-image/compare/v1.1.2...v1.1.3) (2021-07-13)

## [1.1.2](https://github.com/theKashey/holistic-image/compare/v1.1.1...v1.1.2) (2021-07-13)

## [1.1.1](https://github.com/theKashey/holistic-image/compare/v1.1.0...v1.1.1) (2021-07-13)

# [1.1.0](https://github.com/theKashey/holistic-image/compare/v1.0.0...v1.1.0) (2021-07-13)

# 1.0.0 (2021-07-11)


src/constants.ts
import { TargetFormat } from './types';

// derived from https://github.com/GoogleChromeLabs/squoosh/blob/61de471e52147ecdc8ff674f3fcd3bbf69bb214a/libsquoosh/src/codecs.ts
export const defaultConverters: Record<string, TargetFormat> = {
  // with default 75
  jpg: { use: 'mozjpeg', options: ({ scale }) => (scale == 1 ? { quality: 80 } : { quality: 70 }) },
  // with default 75
  webp: { use: 'webp', options: ({ scale }) => (scale == 1 ? { quality: 85, method: 6 } : { quality: 75, method: 5 }) },
  // with default 33 (63-30)
  avif: {
    use: 'avif',
    options: ({ scale }) => (scale == 1 ? { cqLevel: 63 - 43, effort: 5 } : { cqLevel: 63 - 35, effort: 5 }),
  },
};

export const formats = {
  base: ['.jpg', '.png'],
  webp: ['.webp'],
  avif: ['.avif'],
};

// allow 1 second difference between original and derived file
export const MTIME_COMPARE_DELTA = 1000;

export const sizes = ['@1x', '@2x'];

export const DERIVED_PREFIX = 'derived.';
export const HOLISTIC_SIGNATURE = '.holistic.';
export const HOLISTIC_FOLDER = '.holistic';
